tiledb/sm/cpp_api/ndrectangle.h

```diff
@@ -0,0 +1,229 @@
+
+  /* ********************************* */
+  /*                API                */
+  /* ********************************* */
+
+  /**
+   * Adds an 1D range along a dimension name, in the form
+   * (start, end). The datatype of the range
+   * must be the same as the dimension datatype.
+   *
+   * **Example:**
+   *
+   * @code{.cpp}
+   * // Set an 1D range on dimension 0, assuming the domain type is int64.
+   * int64_t start = 10;
+   * int64_t end = 20;
+   * ndrect.set_range(0, start, end);
+   * @endcode
+   *
+   * @tparam T The dimension datatype.
+   * @param dim_name The name of the dimension to add the range to.
+   * @param start The range start to add.
+   * @param end The range end to add.
+   * @return Reference to this NDRectangle.
+   */
+  template <class T>
+  NDRectangle& set_range(const std::string& dim_name, T start, T end) {
+    auto& ctx = ctx_.get();
+
+    // Create the tiledb_range_t struct and fill it
+    tiledb_range_t range;
+    range.min = static_cast<const void*>(&start);
+    range.min_size = sizeof(T);
+    range.max = static_cast<const void*>(&end);
+    range.max_size = sizeof(T);
+
+    // Pass the struct to tiledb_ndrectangle_set_range_for_name
+    ctx.handle_error(tiledb_ndrectangle_set_range_for_name(
+        ctx.ptr().get(), ndrect_.get(), dim_name.c_str(), &range));
+
+    return *this;
+  }
+
+  /**
+   * Adds an 1D range along a dimension index, in the form
+   * (start, end). The datatype of the range
+   * must be the same as the dimension datatype.
+   *
+   * **Example:**
+   *
+   * @code{.cpp}
+   * // Set an 1D range on dimension 0, assuming the domain type is int64.
+   * int64_t start = 10;
+   * int64_t end = 20;
+   * ndrect.set_range(0, start, end);
+   * @endcode
+   *
+   * @tparam T The dimension datatype.
+   * @param dim_idx The index of the dimension to add the range to.
+   * @param start The range start to add.
+   * @param end The range end to add.
+   * @return Reference to this NDRectangle.
+   */
+  template <class T>
+  NDRectangle& set_range(uint32_t dim_idx, T start, T end) {
+    auto& ctx = ctx_.get();
+    // Create the tiledb_range_t struct and fill it
+    tiledb_range_t range;
+    range.min = static_cast<const void*>(&start);
+    // tiledb_datatype_size()
+    range.min_size = sizeof(T);
+    range.max = static_cast<const void*>(&end);
+    range.max_size = sizeof(T);
+
+    // Pass the struct to tiledb_ndrectangle_set_range_for_name
+    ctx.handle_error(tiledb_ndrectangle_set_range(
+        ctx.ptr().get(), ndrect_.get(), dim_idx, &range));
+
+    return *this;
+  }
+
+  /**
+   * Retrieves a range for a given dimension name.
+   * The template datatype must be the same as that of the
+   * underlying array.
+   *
+   * @tparam T The dimension datatype.
+   * @param dim_name The dimension name.
+   * @return A duplex of the form (start, end).
+   */
+  template <class T>
+  std::array<T, 2> range(const std::string& dim_name) {
```

Comment:
We could switch this to a pair?

Reply:
Followed the spec. This is similar to what we do in the subarray class

Reply:
Yeah that's fine for now I guess.